engine/src/main/java/org/terasology/rendering/opengl/AbstractFBOsManager.java

```diff
@@ -0,0 +1,256 @@
+
+/**
+ * TODO: fix above, add javadocs
+ */
+public abstract class AbstractFBOsManager implements BaseFBOsManager {
+    protected static final Logger logger = LoggerFactory.getLogger(AbstractFBOsManager.class);
+    protected Map<ResourceUrn, FBOConfig> fboConfigs = Maps.newHashMap();
+    protected Map<ResourceUrn, FBO> fboLookup = Maps.newHashMap();
+    protected Map<ResourceUrn, Integer> fboUsageCountMap = Maps.newHashMap();
+
+    private Set<FBOManagerSubscriber> fboManagerSubscribers = Sets.newHashSet();
+
+    protected FBO generateWithDimensions(FBOConfig fboConfig, FBO.Dimensions dimensions) {
+        fboConfig.setDimensions(dimensions);
+        FBO fbo = FBO.create(fboConfig);
+
+        // At this stage it's unclear what should be done in this circumstances as I (manu3d) do not know what
+        // the effects of using an incomplete FrameBuffer are. Throw an exception? Live with visual artifacts?
+        if (fbo.getStatus() == FBO.Status.INCOMPLETE) {
+            logger.error("FBO " + fboConfig.getName() + " is incomplete. Look earlier in the log for details.");
+        } else if (fbo.getStatus() == FBO.Status.UNEXPECTED) {
+            logger.error("FBO " + fboConfig.getName() + " has generated an unexpected status code. Look earlier in the log for details.");
+        }
+        fboLookup.put(fboConfig.getName(), fbo);
+        fboConfigs.put(fboConfig.getName(), fboConfig);
+        return fbo;
+    }
+
+    protected void notifySubscribers() {
+        for (FBOManagerSubscriber subscriber : fboManagerSubscribers) {
+            subscriber.update();
+        }
+    }
+
+
+    protected void retain(ResourceUrn resourceUrn) {
+        if (fboUsageCountMap.containsKey(resourceUrn)) {
+            int usageCount = fboUsageCountMap.get(resourceUrn) + 1;
+            fboUsageCountMap.put(resourceUrn, usageCount);
+        } else {
+            fboUsageCountMap.put(resourceUrn, 1);
+        }
+    }
+
+    /**
+     * TODO: add javadoc
+     *
+     * @param fboName
+     */
+    @Override
+    public void release(ResourceUrn fboName) {
+        Preconditions.checkArgument(fboUsageCountMap.containsKey(fboName), "The given fbo is not used.");
+
+        if (fboUsageCountMap.get(fboName) != 1) {
+            int usageCount = fboUsageCountMap.get(fboName);
+            fboUsageCountMap.put(fboName, usageCount - 1);
+        } else {
+            get(fboName).dispose();
+            fboLookup.remove(fboName);
+            if (fboConfigs.containsKey(fboName)) {
+                fboConfigs.remove(fboName);
+            }
+        }
+    }
+
+    /**
+     * Binds the color texture of the FBO with the given name and returns true.
+     *
+     * If no FBO is associated with the given name, false is returned and an error is logged.
+     *
+     * @param fboName the urn of an FBO
+     * @return True if an FBO associated with the given name exists. False otherwise.
+     */
+    public boolean bindFboColorTexture(ResourceUrn fboName) {
```

Comment:
Let's add:

TODO: remove bindFbo*Texture methods from this class and encapsulate the associated methods in the FBO class within StateChange implementations.